website/docs/concepts/router.md

```diff
@@ -192,27 +181,174 @@ You can access them using the following hooks:
+and won't trigger a re-render of the element.
+
+In other words, only use `Callback::once` when you are sure the target route is different. Or use normal callbacks only
+to be safe.
+:::
+
+If you want to replace the current history instead of pushing a new history onto the stack, use `history.replace()`
+instead of `history.push()`.
+
+You may notice `history` has to move into the callback, so it can't be used again for other callbacks. Luckily `history`
+implements `Clone`, here's for example how to have multiple buttons to different routes:
+
+```rust ,ignore
+use yew::prelude::*;
+use yew_router::prelude::*;
+
+#[function_component(NavItems)]
+pub fn nav_items() -> Html {
+    let history = use_history().unwrap();
+
+    let go_home_button = {
+        let history = history.clone();
+        let onclick = Callback::once(move |_| history.push(Route::Home));
+        html! {
+            <button {onclick}>{"click to go home"}</button>
+        }
+    };
+
+    let go_to_first_post_button = {
+        let history = history.clone();
+        let onclick = Callback::once(move |_| history.push(Route::Post { id: "first-post".to_string() }));
+        html! {
+            <button {onclick}>{"click to go the first post"}</button>
+        }
+    };
+
+    let go_to_secure_button = {
+        let onclick = Callback::once(move |_| history.push(Route::Secure));
+        html! {
+            <button {onclick}>{"click to go to secure"}</button>
+        }
+    };
+
+    html! {
+        <>
+            {go_home_button}
+            {go_to_first_post_button}
+            {go_to_secure_button}
+        </>
+    }
+}
+```
+
+##### Struct Components
+
+For struct components, the `AnyHistory` instance can be obtained through the `ctx.link().history()` API. The rest is
+identical with the function component case. Here's an example of a view function that renders a single button.
+
+```rust ,ignore
+fn view(&self, ctx: &Context<Self>) -> Html {
+    let history = ctx.link().history().unwrap();
+    let onclick = Callback::once(move |_| history.push(MainRoute::Home));
+    html!{
+        <button {onclick}>{"Go Home"}</button>
+    }
+}
+```
+
+#### Redirect
+
+`yew-router` also provides a `<Redirect/>` element in the prelude. It can be used to achieve similar effects as the
+history API. The element accepts a
+`to` attribute as the target route. When a `<Redirect/>` element is rendered, it internally calls `history.push()` and
+changes the route. Here is an example:
+
+```rust ,ignore
+#[function_component(SomePage)]
+fn some_page() -> Html {
+    // made-up hook `use_user`
+    let user = match use_user() {
+        Some(user) => user,
+        // an early return that redirects to the login page
+        // technicality: `Redirect` actually renders an empty html. But since it also pushes history, the target page
+        // shows up immediately. Consider it a "side-effect" component.
+        None => return html! { 
+            <Redirect to={Route::Login}/> 
```

Comment:
Should be `<Redirect<Route> to={Route::Login} />`?

Reply:
Should it? If the test doesn't fail, it means type inference works (surprising since it doesn't work in this case)

Reply:
I think the reason is that it accepts a `IntoPropValue<R>`, not directly `R` which caused Rust cannot infer the type.

Reply:
sharp eyes

Reply:
> Should it? If the test doesn't fail, it means type inference works (surprising since it doesn't work in this case)

This snippet is marked as ignore too. there is a madeup `use_user` hook, and for the sake of not repeating a Route struct.